File: 1188-brace-expansion-ii/brace-expansion-ii.cpp
#include <vector>
#include <string>
#include <unordered_set>
#include <algorithm>

using namespace std;

class Solution {
private:
    int ptr = 0;

    unordered_set<string> multiplyFactors(const vector<unordered_set<string>>& factors) {
        if (factors.empty()) return {};

        unordered_set<string> result = {""};
        for (const auto& factor : factors) {
            unordered_set<string> nextResult;
            for (const string& s1 : result) {
                for (const string& s2 : factor) {
                    nextResult.insert(s1 + s2);
                }
            }
            result = move(nextResult);
        }
        return result;
    }

    unordered_set<string> parseExpr(const string& s) {
        unordered_set<string> res;
        vector<unordered_set<string>> currentTermFactors;

        while (ptr < s.length() && s[ptr] != '}') {
            if (s[ptr] == ',') {
                auto combined = multiplyFactors(currentTermFactors);
                res.insert(combined.begin(), combined.end());
                currentTermFactors.clear();
                ptr++; // Skip ','
            } else {
                currentTermFactors.push_back(parseFactor(s));
            }
        }

        auto combined = multiplyFactors(currentTermFactors);
        res.insert(combined.begin(), combined.end());
        return res;
    }

    unordered_set<string> parseFactor(const string& s) {
        if (s[ptr] == '{') {
            ptr++; // Skip '{'
            auto res = parseExpr(s);
            ptr++; // Skip '}'
            return res;
        } else {
            unordered_set<string> res = {string(1, s[ptr])};
            ptr++; // Skip single character
            return res;
        }
    }

public:
    vector<string> braceExpansionII(string expression) {
        ptr = 0;
        auto setRes = parseExpr(expression);
        vector<string> result(setRes.begin(), setRes.end());
        sort(result.begin(), result.end()); // Return sorted unique words
        return result;
    }
};

File: 1188-brace-expansion-ii/brace-expansion-ii.java
import java.util.*;

class Solution {
    private int ptr = 0;

    public List<String> braceExpansionII(String expression) {
        ptr = 0;
        Set<String> set = parseExpr(expression);
        List<String> result = new ArrayList<>(set);
        Collections.sort(result); // Return sorted unique words
        return result;
    }

    private Set<String> parseExpr(String s) {
        Set<String> res = new HashSet<>();
        List<Set<String>> currentTermFactors = new ArrayList<>();

        while (ptr < s.length() && s.charAt(ptr) != '}') {
            if (s.charAt(ptr) == ',') {
                res.addAll(multiplyFactors(currentTermFactors));
                currentTermFactors.clear();
                ptr++; // Skip ','
            } else {
                currentTermFactors.add(parseFactor(s));
            }
        }

        res.addAll(multiplyFactors(currentTermFactors));
        return res;
    }

    private Set<String> parseFactor(String s) {
        if (s.charAt(ptr) == '{') {
            ptr++; // Skip '{'
            Set<String> res = parseExpr(s);
            ptr++; // Skip '}'
            return res;
        } else {
            Set<String> res = new HashSet<>();
            res.add(String.valueOf(s.charAt(ptr)));
            ptr++; // Skip single character
            return res;
        }
    }

    private Set<String> multiplyFactors(List<Set<String>> factors) {
        if (factors.isEmpty()) return Collections.emptySet();

        Set<String> result = new HashSet<>();
        result.add("");

        for (Set<String> factor : factors) {
            Set<String> nextResult = new HashSet<>();
            for (String s1 : result) {
                for (String s2 : factor) {
                    nextResult.add(s1 + s2);
                }
            }
            result = nextResult;
        }

        return result;
    }
}